HeterogeneousCore/SonicTriton/plugins/converters/FloatApFixed16Converter.cc

```diff
@@ -0,0 +1,32 @@
+#include "HeterogeneousCore/SonicTriton/interface/TritonConverterBase.h"
+
+#include <string>
+#include "ap_fixed.h"
+
+template <int I>
+class FloatApFixed16Converter : public TritonConverterBase<float> {
+public:
+  FloatApFixed16Converter() : TritonConverterBase<float>("FloatApFixed16F"+std::to_string(I)+"Converter", 2) {}
+
+  const uint8_t* convertIn(const float* in) const {
+    return reinterpret_cast<const uint8_t*>((this->makeVecIn(in)).data());
+  }
+  const float* convertOut(const uint8_t* in) const {
+    return (this->makeVecOut(reinterpret_cast<const ap_fixed<16, I>*>(in))).data();
```

Comment:
I think this will encounter a lifetime issue. The object returned by `makeVecOut()` goes out of scope as soon as this function returns, so the pointer to its data could point to deallocated or reallocated memory.

In `TritonData`, I added a class member `std::any holder_` to extend the lifetime of the `TritonInput` object in `toServer()`. Something similar is probably needed here. It doesn't need to be `std::any` since the type is known, but it does need to be aware that the `convertIn/Out()` functions can be called multiple times per event (so all created vectors should be stored).

This implies it would be useful for `TritonConverterBase` to have a `clear()` function that can be called in `TritonData::reset()`. To handle the type erasure, it should be stored as an `std::function` member variable in `TritonData`, using `std::bind` when the converter is created.

Reply:
So I have tried to set this up, but I think there may be an issue due to the fact that the converter base class is virtual, and the way that `std::function`/`std::bind` work. If I try to bid a non-virtual function I can get things to work, but since the clear function needs to be virtual (I think) I can't figure out a way to make it work. Perhaps I'm missing something about `std::bind`? I just wanted to comment here in case you have any suggestions, otherwise I will keep playing around.

Reply:
As I understand it, storing a virtual member function in `std::function` should work if you give `std::bind` the pointer to the actual object (even if it's a base class pointer).

This test code worked for me:
```cpp
#include <iostream>
#include <functional>

class Base {
public:
    virtual void doIt() const { std::cout << "Base class" << std::endl; }
};

class Derived: public Base {
public:
    void doIt() const override { std::cout << "Derived class" << std::endl; }
};

int main(){
    Base* obj = new Derived();

    std::function<void()> f = std::bind(&Base::doIt, obj);
    f();
}
```

Reply:
Thanks, I have tried this construction and in a few different attempts to implement it I always get a seg fault. Ill keep trying to work through it

Reply:
I think it should only segfault if the object pointer gets lost/deallocated...

Reply:
In the simplest case I use: `converter_clear = std::bind(&TritonConverterBase<DT>::clear, std::any_cast<std::shared_ptr<TritonConverterBase<DT>>>(converter_));` inside `createConverter()`. I can't really see a way that `converter_` should get lost, and indeed even trying to call `converter_clear()` right after I set it using `std::bind` gives the fault. Perhaps Im missing something obvious here?

Reply:
the output of `std::any_cast<std::shared_ptr<TritonConverterBase<DT>>>(converter_))` has type `std::shared_ptr<TritonConverterBase<DT>`, which is not `TritonConverterBase<DT>*`. Try using `std::any_cast<std::shared_ptr<TritonConverterBase<DT>>>(converter_).get()` instead

Reply:
I have tried both with and without the `.get()`, but both give the fault

Reply:
Can you post the trace from the segfault?

Reply:
`A fatal system signal has occurred: segmentation violation
The following is the call stack containing the origin of the signal.

Mon Nov 30 13:33:25 CST 2020
Thread 7 (Thread 0x7f586a3ff700 (LWP 22023)):
#0  0x00007f588bff6a35 in pthread_cond_wait@@GLIBC_2.3.2 () from /lib64/libpthread.so.0
#1  0x00007f586d70e6e2 in gpr_cv_wait () from /cvmfs/cms.cern.ch/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_0_pre9/external/slc7_amd64_gcc820/lib/libgpr.so.9
#2  0x00007f586fe21a2f in timer_thread(void*) () from /cvmfs/cms.cern.ch/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_0_pre9/external/slc7_amd64_gcc820/lib/libgrpc.so.9
#3  0x00007f586d710823 in grpc_core::(anonymous namespace)::ThreadInternalsPosix::ThreadInternalsPosix(char const*, void (*)(void*), void*, bool*, grpc_core::Thread::Options const&)::{lambda(void*)#1}::_FUN(void*) () from /cvmfs/cms.cern.ch/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_0_pre9/external/slc7_amd64_gcc820/lib/libgpr.so.9
#4  0x00007f588bff2ea5 in start_thread () from /lib64/libpthread.so.0
#5  0x00007f588bd1b96d in clone () from /lib64/libc.so.6
Thread 6 (Thread 0x7f586b1ff700 (LWP 22017)):
#0  0x00007f588bff6de2 in pthread_cond_timedwait@@GLIBC_2.3.2 () from /lib64/libpthread.so.0
#1  0x00007f586d70e74a in gpr_cv_wait () from /cvmfs/cms.cern.ch/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_0_pre9/external/slc7_amd64_gcc820/lib/libgpr.so.9
#2  0x00007f586fe21a2f in timer_thread(void*) () from /cvmfs/cms.cern.ch/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_0_pre9/external/slc7_amd64_gcc820/lib/libgrpc.so.9
#3  0x00007f586d710823 in grpc_core::(anonymous namespace)::ThreadInternalsPosix::ThreadInternalsPosix(char const*, void (*)(void*), void*, bool*, grpc_core::Thread::Options const&)::{lambda(void*)#1}::_FUN(void*) () from /cvmfs/cms.cern.ch/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_0_pre9/external/slc7_amd64_gcc820/lib/libgpr.so.9
#4  0x00007f588bff2ea5 in start_thread () from /lib64/libpthread.so.0
#5  0x00007f588bd1b96d in clone () from /lib64/libc.so.6
Thread 5 (Thread 0x7f586c3dc700 (LWP 22016)):
#0  0x00007f588bff6a35 in pthread_cond_wait@@GLIBC_2.3.2 () from /lib64/libpthread.so.0
#1  0x00007f586d70e6e2 in gpr_cv_wait () from /cvmfs/cms.cern.ch/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_0_pre9/external/slc7_amd64_gcc820/lib/libgpr.so.9
#2  0x00007f586fe0e6fb in grpc_core::Executor::ThreadMain(void*) () from /cvmfs/cms.cern.ch/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_0_pre9/external/slc7_amd64_gcc820/lib/libgrpc.so.9
#3  0x00007f586d710823 in grpc_core::(anonymous namespace)::ThreadInternalsPosix::ThreadInternalsPosix(char const*, void (*)(void*), void*, bool*, grpc_core::Thread::Options const&)::{lambda(void*)#1}::_FUN(void*) () from /cvmfs/cms.cern.ch/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_0_pre9/external/slc7_amd64_gcc820/lib/libgpr.so.9
#4  0x00007f588bff2ea5 in start_thread () from /lib64/libpthread.so.0
#5  0x00007f588bd1b96d in clone () from /lib64/libc.so.6
Thread 4 (Thread 0x7f586cbdd700 (LWP 22015)):
#0  0x00007f588bff6a35 in pthread_cond_wait@@GLIBC_2.3.2 () from /lib64/libpthread.so.0
#1  0x00007f586d70e6e2 in gpr_cv_wait () from /cvmfs/cms.cern.ch/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_0_pre9/external/slc7_amd64_gcc820/lib/libgpr.so.9
#2  0x00007f586fe0e6fb in grpc_core::Executor::ThreadMain(void*) () from /cvmfs/cms.cern.ch/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_0_pre9/external/slc7_amd64_gcc820/lib/libgrpc.so.9
#3  0x00007f586d710823 in grpc_core::(anonymous namespace)::ThreadInternalsPosix::ThreadInternalsPosix(char const*, void (*)(void*), void*, bool*, grpc_core::Thread::Options const&)::{lambda(void*)#1}::_FUN(void*) () from /cvmfs/cms.cern.ch/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_0_pre9/external/slc7_amd64_gcc820/lib/libgpr.so.9
#4  0x00007f588bff2ea5 in start_thread () from /lib64/libpthread.so.0
#5  0x00007f588bd1b96d in clone () from /lib64/libc.so.6
Thread 3 (Thread 0x7f586d3de700 (LWP 22014)):
#0  0x00007f588bff6a35 in pthread_cond_wait@@GLIBC_2.3.2 () from /lib64/libpthread.so.0
#1  0x00007f588c5d962c in __gthread_cond_wait (__mutex=<optimized out>, __cond=<optimized out>) at /data/cmsbld/jenkins/workspace/auto-builds/CMSSW_11_1_0_pre7-slc7_amd64_gcc820/build/CMSSW_11_1_0_pre7-build/BUILD/slc7_amd64_gcc820/external/gcc/8.2.0-bcolbf/gcc-8.2.0/obj/x86_64-unknown-linux-gnu/libstdc++-v3/include/x86_64-unknown-linux-gnu/bits/gthr-default.h:864
#2  std::condition_variable::wait (this=<optimized out>, __lock=...) at ../../../../../libstdc++-v3/src/c++11/condition_variable.cc:53
#3  0x00007f58701cb277 in SonicDispatcherPseudoAsync::waitForNext() () from /cvmfs/cms.cern.ch/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_0_pre9/lib/slc7_amd64_gcc820/libHeterogeneousCoreSonicCore.so
#4  0x00007f588c5ded2f in std::execute_native_thread_routine (__p=0x7f587132d9a0) at ../../../../../libstdc++-v3/src/c++11/thread.cc:80
#5  0x00007f588bff2ea5 in start_thread () from /lib64/libpthread.so.0
#6  0x00007f588bd1b96d in clone () from /lib64/libc.so.6
Thread 2 (Thread 0x7f5870e37700 (LWP 22003)):
#0  0x00007f588bffa1d9 in waitpid () from /lib64/libpthread.so.0
#1  0x00007f5886d6afd7 in edm::service::cmssw_stacktrace_fork() () from /cvmfs/cms.cern.ch/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_0_pre9/lib/slc7_amd64_gcc820/pluginFWCoreServicesPlugins.so
#2  0x00007f5886d6ba9a in edm::service::InitRootHandlers::stacktraceHelperThread() () from /cvmfs/cms.cern.ch/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_0_pre9/lib/slc7_amd64_gcc820/pluginFWCoreServicesPlugins.so
#3  0x00007f588c5ded2f in std::execute_native_thread_routine (__p=0x7f5881b6f1b0) at ../../../../../libstdc++-v3/src/c++11/thread.cc:80
#4  0x00007f588bff2ea5 in start_thread () from /lib64/libpthread.so.0
#5  0x00007f588bd1b96d in clone () from /lib64/libc.so.6
Thread 1 (Thread 0x7f588a304540 (LWP 21976)):
#0  0x00007f588bd10c3d in poll () from /lib64/libc.so.6
#1  0x00007f5886d6b43f in full_read.constprop () from /cvmfs/cms.cern.ch/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_0_pre9/lib/slc7_amd64_gcc820/pluginFWCoreServicesPlugins.so
#2  0x00007f5886d6bb7c in edm::service::InitRootHandlers::stacktraceFromThread() () from /cvmfs/cms.cern.ch/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_0_pre9/lib/slc7_amd64_gcc820/pluginFWCoreServicesPlugins.so
#3  0x00007f5886d6ca59 in sig_dostack_then_abort () from /cvmfs/cms.cern.ch/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_0_pre9/lib/slc7_amd64_gcc820/pluginFWCoreServicesPlugins.so
#4  <signal handler called>
#5  0x0000000000000000 in ?? ()
#6  0x00007f58701f2264 in void TritonData<nvidia::inferenceserver::client::InferInput>::toServer<float>(std::shared_ptr<std::vector<std::vector<float, std::allocator<float> >, std::allocator<std::vector<float, std::allocator<float> > > > >) () from /uscms_data/d3/drankin/Sonic/CMSSW_11_2_0_pre9/lib/slc7_amd64_gcc820/libHeterogeneousCoreSonicTriton.so
#7  0x00007f5870215259 in TritonImageProducer::acquire(edm::Event const&, edm::EventSetup const&, std::unordered_map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, TritonData<nvidia::inferenceserver::client::InferInput>, std::hash<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::equal_to<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, TritonData<nvidia::inferenceserver::client::InferInput> > > >&) () from /uscms_data/d3/drankin/Sonic/CMSSW_11_2_0_pre9/lib/slc7_amd64_gcc820/plugintestHeterogenousCoreSonicTriton.so
#8  0x00007f587020f55b in SonicAcquirer<TritonClient, edm::stream::EDProducer<edm::ExternalWork> >::acquire(edm::Event const&, edm::EventSetup const&, edm::WaitingTaskWithArenaHolder) () from /uscms_data/d3/drankin/Sonic/CMSSW_11_2_0_pre9/lib/slc7_amd64_gcc820/plugintestHeterogenousCoreSonicTriton.so
#9  0x00007f588e8bb319 in edm::stream::doAcquireIfNeeded(edm::stream::impl::ExternalWork*, edm::Event const&, edm::EventSetup const&, edm::WaitingTaskWithArenaHolder&) () from /cvmfs/cms.cern.ch/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_0_pre9/lib/slc7_amd64_gcc820/libFWCoreFramework.so
#10 0x00007f588e8b994f in edm::stream::EDProducerAdaptorBase::doAcquire(edm::EventTransitionInfo const&, edm::ActivityRegistry*, edm::ModuleCallingContext const*, edm::WaitingTaskWithArenaHolder&) () from /cvmfs/cms.cern.ch/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_0_pre9/lib/slc7_amd64_gcc820/libFWCoreFramework.so
#11 0x00007f588e886a62 in edm::Worker::runAcquire(edm::EventTransitionInfo const&, edm::ParentContext const&, edm::WaitingTaskWithArenaHolder&) () from /cvmfs/cms.cern.ch/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_0_pre9/lib/slc7_amd64_gcc820/libFWCoreFramework.so
#12 0x00007f588e886bc1 in edm::Worker::runAcquireAfterAsyncPrefetch(std::__exception_ptr::exception_ptr const*, edm::EventTransitionInfo const&, edm::ParentContext const&, edm::WaitingTaskWithArenaHolder) () from /cvmfs/cms.cern.ch/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_0_pre9/lib/slc7_amd64_gcc820/libFWCoreFramework.so
#13 0x00007f588e7fafb3 in edm::Worker::AcquireTask<edm::OccurrenceTraits<edm::EventPrincipal, (edm::BranchActionType)1>, void>::execute() () from /cvmfs/cms.cern.ch/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_0_pre9/lib/slc7_amd64_gcc820/libFWCoreFramework.so
#14 0x00007f588cfecbfd in tbb::internal::custom_scheduler<tbb::internal::IntelSchedulerTraits>::process_bypass_loop (this=this@entry=0x7f58895b3200, context_guard=..., t=t@entry=0x7f58895a7340, isolation=isolation@entry=0) at ../../src/tbb/custom_scheduler.h:393
#15 0x00007f588cfecef5 in tbb::internal::custom_scheduler<tbb::internal::IntelSchedulerTraits>::local_wait_for_all (this=0x7f58895b3200, parent=..., child=<optimized out>) at ../../include/tbb/task.h:1003
#16 0x00007f588e77b355 in edm::EventProcessor::processLumis(std::shared_ptr<void> const&) () from /cvmfs/cms.cern.ch/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_0_pre9/lib/slc7_amd64_gcc820/libFWCoreFramework.so
#17 0x00007f588e78349e in edm::EventProcessor::runToCompletion() () from /cvmfs/cms.cern.ch/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_0_pre9/lib/slc7_amd64_gcc820/libFWCoreFramework.so
#18 0x000000000040f59d in tbb::interface7::internal::delegated_function<main::{lambda()#1}::operator()() const::{lambda()#1} const, void>::operator()() const ()
#19 0x00007f588cfe7bc1 in tbb::interface7::internal::task_arena_base::internal_execute (this=0x7ffdbfdfaf50, d=...) at ../../src/tbb/arena.cpp:1105
#20 0x0000000000410488 in main::{lambda()#1}::operator()() const ()
#21 0x000000000040eff5 in main ()

Current Modules:

Module: TritonImageProducer:TritonProducer (crashed)

A fatal system signal has occurred: segmentation violation`

Reply:
A clean build attempt seems to fix this issue, so the `clear()` is now implemented